### PR TITLE
#804 Update Allowed Tags And Attributes 

### LINKED ISSUE #761: Playbuzz not working in 0.5.1
http://www.playbuzz.com/

### LINKED ISSUE #772: A not allowed script tag appeared
Hey guys, after updating my plugin Im with some issues according to Google Search Console.

AMP Page for example: https://todateen.com.br/anitta-dj-alok-produzindo-parceria/amp/

The Amp Validator (https://validator.ampproject.org/) says that is a <script> tag not allowed
Printscreen: https://sketch.io/render/sk-974a93e9b0d78d4a3ac7bc2e9a57ba73.jpeg

The playbuzz tag:` [playbuzz-item item="5ea7b9e0-d919-4de0-b729-43171f0a2729" info="false" shares="false" comments="false" recommend="undefined" format="video-snaps"]`

Can you help me, please? :)

### LINKED ISSUE #804: Update allowed tags according to AMP specification
**[Edited] Refer to the latest [acceptance criteria](https://github.com/Automattic/amp-wp/issues/798#issuecomment-346595030).**
___
It appears that the [`amp_wp_build.py`](https://github.com/Automattic/amp-wp/blob/647b003d90d7141ca7b54329ca186942ae149e46/bin/amp_wp_build.py) script hasn't been run since it was introduced at the end of 2016. Since that time, the [changes](https://github.com/Automattic/amp-wp/commits/master/includes/sanitizers/class-amp-allowed-tags-generated.php) to [`class-amp-allowed-tags-generated.php`](https://github.com/Automattic/amp-wp/blob/647b003d90d7141ca7b54329ca186942ae149e46/includes/sanitizers/class-amp-allowed-tags-generated.php) appear to have been done manually. There have been [95 commits](https://github.com/ampproject/amphtml/commits/master?path%5B%5D=spec) to the `spec` since the end of 2016, and these changes need to be comprehensively added to the plugin. 

I understand the `amp_wp_build.py` script has to be updated to account for some changes to the spec and to the amphtml project. For example, a dependency has been removed which has to be reverted or restored in some way: https://github.com/ampproject/amphtml/pull/10830

Additionally, the spec changes need to be more regularly applied, such as via Travis CI cron, if not done manually.

### PR DESCRIPTION
**Pull request by kienstra**
This PR fixes and improves the AMP HTML tag and attributes updater, [refer to the ACs](https://github.com/Automattic/amp-wp/issues/804#issuecomment-346622935) for more info...

___

**Steps To Manually Run Script**

In order for `amp_wp_build.py` to generate `class-amp-allowed-tags-generated.php`, run these commands, using a Linux environment like [VVV](https://github.com/Varying-Vagrant-Vagrants/VVV):
1. `cd /path/to/your/vvv/wp-content/plugins/amp-wp`
2. `bash bin/amphtml-update.sh`

**Note:** these steps were edited, as most of these steps are now in the file `bin/amphtml-update.sh`. That file also includes the step:
7. `git apply` this [commit](https://drive.google.com/file/d/1DntvrqjGmgciPEvxkWG6KBn0FycXFVqk/view?usp=sharing) to re-add `validator_gen_md.py`, which this [commit](https://github.com/ampproject/amphtml/pull/10830/commits) deleted

See #689 (Only the requirement "Minor: The script should do the git clone...")
Fixes #761
Fixes #772
Fixes #804 
Closes #749
Closes #805

### CODE REVIEW

File: includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php [hunk 1]
```diff
@@ -138,24 +138,18 @@ private function process_node( $node ) {
 				// This should not happen very often, but...
 				// If we're here, then we're not sure which spec should
 				// be used. Let's use the top scoring ones.
-				foreach( $spec_ids_sorted as $id ) {
-					$attr_spec_list = array_merge( $attr_spec_list, $rule_spec_list_to_validate[ $id ][AMP_Rule_Spec::ATTR_SPEC_LIST] );
+				foreach ( $spec_ids_sorted as $id ) {
+					$spec_list = isset( $rule_spec_list_to_validate[ $id ][ AMP_Rule_Spec::ATTR_SPEC_LIST ] ) ? $rule_spec_list_to_validate[ $id ][ AMP_Rule_Spec::ATTR_SPEC_LIST ] : null;
+					if ( ! $this->is_missing_mandatory_attribute( $spec_list, $node ) ) {
```
Comment (kienstra) [inline includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php]: This addresses an edge case where there are more than one spec lists, and one of them requires a value. If one of them doesn't require a value, use that instead. 

This detects that with `$this->is_missing_mandatory_attribute()`. If the node if it satisfies one of the specs, this doesn't remove it.

File: includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php [hunk 2]
```diff
@@ -138,24 +138,18 @@ private function process_node( $node ) {
...
-				isset( $attr_spec_rule_value[ AMP_Rule_Spec::MANDATORY ] )
-				? (bool) $attr_spec_rule_value[ AMP_Rule_Spec::MANDATORY ]
-				: false;
-			$attribute_exists = $node->hasAttribute( $attr_name );
+		} // End if().
 
-			if ( $is_mandatory && ! $attribute_exists ) {
-				$this->remove_node( $node );
-				return;
-			}
```
Comment (kienstra) [inline includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php]: This is mainly decomposed into the function below, `is_missing_mandatory_attribute()`.

File: includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php [hunk 3]
```diff
@@ -430,8 +449,7 @@ private function _sanitize_disallowed_attribute_values_in_node( $node, $attr_spe
 		foreach ( $attrs_to_remove as $attr_name ) {
 			if ( isset( $attr_spec_list[$attr_name][AMP_Rule_Spec::ALLOW_EMPTY] ) &&
 				( true == $attr_spec_list[$attr_name][AMP_Rule_Spec::ALLOW_EMPTY] ) ) {
-				$attr = $node->attributes;
-				$attr[ $attr_name ]->value = '';
+				$node->setAttribute( $attr_name, '' );
```
Comment (kienstra) [inline includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php]: This raised an error, so it uses the `setAttribute()` method instead.

File: phpcs.xml [hunk 4]
```diff
@@ -19,4 +19,5 @@
 	<exclude-pattern>*/dev-lib/*</exclude-pattern>
 	<exclude-pattern>*/node_modules/*</exclude-pattern>
 	<exclude-pattern>*/vendor/*</exclude-pattern>
+	<exclude-pattern>*/includes/sanitizers/class-amp-allowed-tags-generated.php</exclude-pattern>
```
Comment (kienstra) [inline phpcs.xml]: This file is generated by the Python file, and created hundreds of warnings and errors in PHPCS.

File: tests/amp-tag-and-attribute-sanitizer-private-methods-tests.php [hunk 5]
```diff
@@ -246,136 +246,6 @@ public function get_attr_spec_rule_data() {
 				),
 				'expected' => AMP_Rule_Spec::NOT_APPLICABLE,
 			),
-
```
Comment (kienstra) [inline tests/amp-tag-and-attribute-sanitizer-private-methods-tests.php]: The allowed protocol values are removed from the spec file, so this removes the tests for them.

**Comment by westonruter** [comment]
Double-arrow PHPCS errors are suppressed for now via 2b1443ed25e42e22877d9f6c47d7b633fb14680c.

**Comment by westonruter** [comment]
@kienstra I think the next thing here is to look over gitHub issues that seem to be related to the spec being out of date, and confirm that the issues are now resolved with your latest changes.

File: bin/amphtml-update.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
Comment (ThierryA) [inline bin/amphtml-update.sh]: Here is a draft the bash script to update AMP HTML. This script is meant to run in Linux environment (ex. VVV). To run the script, simply `cd` to the plugin root and run `bash bin/amphtml-update.sh`.

File: bin/amphtml-update.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+set -e
+
+# Go to the right location.
+if [[ '.' != $(dirname "$0") ]]; then
+	cd bin
+fi
```
Comment (westonruter) [inline bin/amphtml-update.sh]: What I usually do instead of this conditional is:

```
cd "$(dirname "$0")"
```

File: bin/amphtml-update.sh [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+if ! apt-get > /dev/null; then
+	echo -e "The AMP HTML uses the apt-get, make sure to run this script in a Linux environment"
+	exit
+fi
+
+# Install dependencies.
+sudo apt-get install git
+sudo apt-get install python
+sudo apt-get install protobuf-compiler
+sudo apt-get install python-protobuf
```
Comment (westonruter) [inline bin/amphtml-update.sh]: You can combine all of the dependencies into one call:

```
sudo apt-get install git python protobuf-compiler python-protobuf
```

File: bin/amphtml-update.sh [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+# Create and go to vendor.
+if [[ ! -e $VENDOR_PATH ]]; then
+	mkdir $VENDOR_PATH
+fi
+cd $VENDOR_PATH
+
+# Clone amphtml repo.
+if [[ ! -e $VENDOR_PATH/amphtml ]]; then
+	git clone https://github.com/ampproject/amphtml amphtml
+fi
```
Comment (westonruter) [inline bin/amphtml-update.sh]: Maybe add an `else` here which goes into `$VENDOR_PATH/amphtml` and does `git pull`?

File: bin/amphtml-update.sh [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+VENDOR_PATH=$PROJECT_PATH/vendor
+
+if ! apt-get > /dev/null; then
+	echo -e "The AMP HTML uses the apt-get, make sure to run this script in a Linux environment"
+	exit
+fi
+
+# Install dependencies.
+sudo apt-get install git
+sudo apt-get install python
```
Comment (westonruter) [inline bin/amphtml-update.sh]: Maybe we should make `git` and `python` just dependencies that we assume are installed, and exit if the command doesn't exist?

```bash
if ! command -v git >/dev/null 2>&1; then
    echo "Git is required"
    exit 1
fi
if ! command -v python >/dev/null 2>&1; then
    echo "Python is required"
    exit 1
fi
```

File: bin/amphtml-update.sh [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+# Go to the right location.
+if [[ '.' != $(dirname "$0") ]]; then
+	cd bin
+fi
+
+BIN_PATH="$(pwd)"
+PROJECT_PATH=$(dirname $PWD)
+VENDOR_PATH=$PROJECT_PATH/vendor
+
+if ! apt-get > /dev/null; then
```
Comment (westonruter) [inline bin/amphtml-update.sh]: Better to check if the command exists via: `if ! command -v apt-get >/dev/null 2>&1; then`

File: bin/amphtml-update.sh [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	echo -e "The AMP HTML uses the apt-get, make sure to run this script in a Linux environment"
+	exit
```
Comment (westonruter) [inline bin/amphtml-update.sh]: Do `exit 1` so a failure error code is sent. An `exit` means just `exit 0` which means success.

File: bin/amphtml-update.sh [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+if [[ '.' != $(dirname "$0") ]]; then
+	cd bin
+fi
+
+BIN_PATH="$(pwd)"
+PROJECT_PATH=$(dirname $PWD)
+VENDOR_PATH=$PROJECT_PATH/vendor
+
+if ! apt-get > /dev/null; then
+	echo -e "The AMP HTML uses the apt-get, make sure to run this script in a Linux environment"
```
Comment (westonruter) [inline bin/amphtml-update.sh]: No need for `-e`

File: bin/amphtml-update.sh (hunk 10 above)
Comment (ThierryA) [reply bin/amphtml-update.sh]: I think it is fine to automatically install `git` or `python` if not installed, the check is pretty quick if it is already installed.

**Comment by ThierryA** [comment]
@kienstra now that the bash script to update the HTML AMP tags is completed, you may update the PR description with the steps to run the script:

1. `cd` to the AMP plugin root
2. run `bash bin/amphtml-update.sh`

Note that the `amphtml-update.sh` bash script is meant to run in Linux environment (ex. VVV).

**Comment by westonruter** [comment]
@kienstra This PR will close PR #749 as obsolete, correct?

**Comment by kienstra** [comment]
**Created `contributing.md`, Updated Script Steps,** 

Hi @westonruter and @ThierryA,
Thanks a lot for your work on this. The commit above adds a [contributing.md](https://github.com/Automattic/amp-wp/blob/eab6207eccc7d1089d05c6b3af4c524bf235d292/contributing.md) file, with instructions to use your new script `amphtml-update.sh`. At some point, that file should probably have more details about contributing. 

Still, I need to look at the remaining [issues](https://github.com/Automattic/amp-wp/issues) for AMP tag support, and see if this PR addresses them.

**Comment by ThierryA** [comment]
@kienstra as you go through the PRs and issues that this PR address, could you add the `Fixes #{number}` and/or `Closes #{number}` to the description (I have already added a few). Thanks,

File: contributing.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+### Updating Allowed Tags And Attributes
+
+The file `class-amp-allowed-tags-generated.php` has the AMP specification's allowed tags and attributes. It's used in sanitization.
+To update that file:
+1. `cd` to the root of this plugin
+2. run `bash bin/amphtml-update.sh`
+That script is intended for a Linux environment like [VVV](https://github.com/Varying-Vagrant-Vagrants/VVV).
+
+### PHPUnit Testing
```
[the remark this answers is not in the document]
Comment (westonruter) [reply contributing.md]: Yes, this came up in #828,.along with questions about how to get pre-commit hook working from wp-dev-lib.

**Comment by kienstra** [comment]
Sure, @ThierryA. I'll add `Fixes #{number}` and or `Closes #{number}` to related tickets, as they apply.

File: contributing.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (kienstra) [reply contributing.md]: Thanks, @ThierryA. This [commit](https://github.com/Automattic/amp-wp/pull/823/commits/44554c543dedffa6c2bf5fa2cb3307c62a6029b8) adds your point to [contributing.md](https://github.com/Automattic/amp-wp/blob/44554c543dedffa6c2bf5fa2cb3307c62a6029b8/contributing.md):
```
Please run these tests in an environment with WordPress unit tests installed, like [VVV](https://github.com/Varying-Vagrant-Vagrants/VVV).
```

File: contributing.md [hunk 15]
```diff
@@ -0,0 +1,30 @@
+# AMP Contributing Guide
+
```
[the remark this answers is not in the document]
Comment (kienstra) [reply contributing.md]: Thanks, @ThierryA. This [commit](https://github.com/Automattic/amp-wp/pull/823/commits/44554c543dedffa6c2bf5fa2cb3307c62a6029b8) adds your sentence to [contributing.md](https://github.com/Automattic/amp-wp/blob/44554c543dedffa6c2bf5fa2cb3307c62a6029b8/contributing.md):
```
Thanks for taking the time to contribute!
```

**Comment by kienstra** [comment]
**Added Issues That Are Fixed**

Hi @ThierryA and @westonruter,
Having looked at all of the issues with the labels [validation-errors](https://github.com/Automattic/amp-wp/labels/validation-errors) and [sanitizers](https://github.com/Automattic/amp-wp/labels/sanitizers), I added these issues to the PR description as fixed by this PR:

#689 (Only the requirement "Minor: The script should do the git clone...")
#761
#772

The following issues also look to have been resolved, though not with this PR. I made a comment to that effect on the ticket, requesting that the openers verify them:
#237 
#253
#781

**Comment by kienstra** [comment]
**Resolved Merge Conflicts**

Hi @westonruter,
Thanks for waiting. This [commit](https://github.com/Automattic/amp-wp/pull/823/commits/c3b28c00b7c8227ab13c23e479198ab3aadd5e7d) merges in `develop`, and resolves the merge conflicts. The commits that follow fix a PHPUnit test and PHPCS issues.

**Pull request merged**